middleware/ratelimiter/rate_limiter.go

```diff
@@ -0,0 +1,78 @@
+package ratelimiter
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/gobuffalo/buffalo"
+)
+
+// Counter limits the amount of times an IP can run
+type Counter interface {
+	Increment(ip string) (count int, err error)
+	Decrement(ip string) (count int, err error)
+	Count(ip string) (count int, err error)
+	Set(ip string, count int) error
+}
+
+// DefaultOptions is the default configuration for Rate Limiter middleware.
+// By default, it fetches the IP based on IPHeaders
+var DefaultOptions = Options{
+	Operations: 100,
+	Duration:   time.Second,
+	GetIP: func(c buffalo.Context) (ip string, ok bool) {
+		req := c.Request()
+		for _, h := range []string{"X-Real-IP", "X-Client-IP", "X-Forwarded-For", "RemoteAddr"} {
+			ip := req.Header.Get(h)
+			if len(ip) > 0 {
+				return ip, true
+			}
+		}
+		return
+	},
+}
+
+// Options define options for rate-limiter middleware
+type Options struct {
+	// Operations is how many operations per Duration an IP is allowed
+	Operations int
+
+	// Duration is how long to count Operations
+	Duration time.Duration
+
+	// GetIP is the func used to get the IP address from the request
+	GetIP func(c buffalo.Context) (ip string, ok bool)
+}
+
+// Middleware is the http middleware that limits the interactions to the API
+// based on the IP address using the provided Counter
+func Middleware(r Counter, opts *Options) buffalo.MiddlewareFunc {
```

Comment:
I didn't have it as a pointer at first, but changed it to allow users to change their options while in flight. If you pass Options at startup but decide to change the limited duration you can adjust it based on some logic without restarting.

Reply:
In that case, it might be better to change it to a function instead. It's very possible that different IPs would have different rate limits. Think someone who pays for API access over someone who doesn't.

Reply:
Makes sense, `getOpts func(ip string) Options` perhaps?

Reply:
Sorry, this slipped to the back of my mind the last couple of days. Thinking about it now there are a couple of things we need to think about. The first, is what we've already been discussing, the other is this implementation is very tied to the idea that the `IP` is how rate limiting will work. What about if I wanted to rate limit by a token, username, email, etc...?

I'm thinking an implementation like this might work more effectively, and give the best flexibility:

```go
func New(r Counter, fn func(c buffalo.Context) Options) buffalo.MiddlewareFunc {
	return func(next buffalo.Handler) buffalo.Handler {
		return func(c buffalo.Context) error {
			opts = fn(c)
			// logic here ...
		}
	}
}
```

This would allow the end user to determine, for each request, how long the duration is, how many hits they can have, how to look up that information (IP, email, etc...)

We should provide a "default" implementation that uses IP.

Thoughts?

Reply:
Yeah this makes perfect sense to me. I started down that path by allowing `func GetIP` to be customized, but it makes more sense to generalize it more yet still provide the IP defaults. I'll work on that soon. Perhaps this evening.

Reply:
Awesome. Also, I think this should go in the `github.com/gobuffalo/x` repo, so I'm going to close this PR, if you wouldn't mind opening another one there. Please make sure to add a `README.md` for it so people know how to use it. Thanks.